README
IperBackup - a Ipernity backup script
=====================================
IperBackup gives you the possibility to completely download all documents in
you Ipernity account. You can either let IperBackup do all the work, so that
it fetches all information and downloads every single file or you let
IperBackup just create a download URL list, which you can use with your
favourite DL manager.

Requirements
============
To use IperBackup your system needs to have the following requirements installed 
(best installed from bottom to top):

	- Perl 5.8.x (or higher)
	- Ipernity::API 0.09 (or higher)
	- Date::Manip
	- XML::Simple
	- LWP::UserAgent
	- URI
	- HTTP::Request
	- Digest::MD5
	- Log::Log4perl
	- Getopt::Long
	- Time::HiRes

CPAN 101
========
To check if the corresponding perl module is installed, you can use the
following command:

	perl -M<module name> -eprint
	e. g.: perl -MIpernity::API -eprint

If this command returns an error message, the module is missing on your
system. If nothing is returned, everything is fine.

To install a module, the easiest way is to use CPAN. CPAN is the
"Comprehensive Perl Archive Network". To install a module you can use the
follwoing command.

	perl -MCPAN -eshell

This will start the CPAN shell for you. Once you see the "cpan>" prompt,
just enter "install <module name>" and CPAN will fetch the module and 
install it locally on your system

Usage
=====
The usage of IperBackup is pretty simple. First you have to decide which action to use:
	
	-l, --list	to create a list with all download URL, to use in your DL manager
	or
	-d, --download to let IperBackup download each file for you
	
	the options are mutual exclusive and mandatory. Otherwise IperBackup
	will just display the help output.

You should also specify an output directory, so IperBackup knows where to store your
files (or the URL list). If you don't specify the output directory, IperBackup will
use /var/tmp. The parameter for this is:

	-o, --outdir

You also will need a config file, where you define some parameters, which are mandatory
to IperBackup (e. g. your API AuthToken and the API UserID). By default IperBackup will
look for the config file in /etc/IperBackup.conf. If you wanna use another file, just 
use the:

	-c, --config

parameter to tell IperBackup where to find the config file.

If you haven't specified your API UserID and AuthToken in the config file, IperBackup 
will generate and fetch both for you, so you can add it to your config file. Once both
parameters are set in your config file, IperBackup won't bother you about this anymore.

If you like to fetch only a specific media type (there are: audio, photo, video and 
other), you can use the -m (or --media) parameter. You can specify just one media type, 
but also you can specify more media types at once. Here are two examples:

        -m video
        or
        --media video,audio,other

In case you like to fetch only a documents with a list of specified tags, you can use
the -t (or --tags) parameter. You can provide one or more tags (up to 20). The search 
will be more limited, the more tags you provide. Here are two usage examples:

        -t 2009
        or
        --tags 2009,cologne,water

You also have to the possiblity to fetch comments for each documents (if there
are some present). Just use the -n (or --comments) parameter and your comments
will be automatically fetched and stored in the same directory as your output
directory (with a trailing _comments.txt added to the original filename). If
you already fetched all your files, but forgot the comments, you can add the 
"--commentsonly" parameter to your download job and it will only fetch the 
comments and store them into the output directory.

If you want to fetch only documents within a specific timeframe, you can use
the following two parameters:

	-s or --startdate <date in UTC>
	and
	-e or --enddate <date in UTC>

The specified date has to be UTC. If you leave the start date it will use the
start of the UNIX epoch time (1970-01-01 00:00:00) and if you leave out the
end date it will use the current time.

When downloading files from your Ipernity account IperBackup will skip files which 
are already present in your local output directory. If you want those files downloaded 
anyhow use the following switch to have IperBackup add a timestamp to the original 
file name and place it in your local directory
	
	-ts or --timestamp

By default IperBackup will fetch the permissions (Ipernity-wise) for each
document and will store it in the filename. The permission string will be
stored in the beginning of the filename. The result will look something
like this: 
	perm_public-<original filename> for public files,
	perm_private-<original filename> for only private,
	perm_familyonly-<original filename> for private + family,
	perm_friendonly-<original filename> for private + friends,
	perm_familyandfriend-<original filename> for private + family + friends.
You can disable this behaviour by using this switch:

	-p or --nopermission

Thus the file will be saved with their original filename.


That's all. Have fun using it!

Comments and critics are always welcome at: doomy@pebcak.de


$Id$


Changelog
IperBackup v0.09 // 2011-01-XX
==============================
- Added possibility to fetch albums
- IperBackup requires Ipernity::API 0.10 now

IperBackup v0.08 // 2011-01-07
==============================
- Default config file path is ./IperBackup.conf instead of 
  /etc/IperBackup.conf (Issue #2)
- Instead of adding a timestamp to already downloaded documents, the file
  will be skipped by default. Timestamps will only be added if the -ts
  switch is given (Issue #3)
- Permission fetching is enabled by default (Issue #4)
- Better documentation. Thanks to Falko Zurell (@maxheadroom) for his work
- IperBackup requires Ipernity::API 0.09 now
- Added --version / -v switch

IperBackup v0.07 // 2011-01-02
==============================
- Added possibilty to fetch document permissions from Ipernity (Requested by @maxheadroom)
  http://iperbackup.my-webspace.org/index.php?p=/discussion/3/is-it-possible-to-retrieve-the-permissions-for-each-document-from-ipernity

IperBackup v0.06 // 2010-12-13
==============================
- Added start-/enddate feature
- "download" mode is now implied when using "commentsonly" parameter
- Code cleanup

IperBackup v0.05 // 2010-12-08
==============================
- Added "commentsonly" feature (requested by Sascha Assbach)
- Added "Update notification" feature. IperBackup will notify you whenever a new release 
  of IperBackup is available
- Creation date of downloaded files, will be changed according to the "creation date" of 
  the picture at Ipernity if available. (requested by Michael Behrens)
- Documentation update
- Added "CPAN 101" to README
- Important: This release requires Ipernity::API 0.08 or higher

IperBackup v0.04 // 2010-11-13
==============================
- Added "comments" download feature

IperBackup v0.03 // 2010-09-28
==============================
- Added possibility to search tags
- Bugfixes in document search
- Important: This release requires Ipernity::API 0.07 or higher

IperBackup v0.02 // 2010-09-27
==============================
- Added possibility to fetch only specific media types
- Documentation update

IperBackup v0.01 // 2010-09-26
==============================
- First official release

$Id$


IperBackup/Config.pm
#!/usr/bin/perl -wt
#
# Filename:     IperBackup/Config.pm
# Description:  Module to handle the configuration file
# Creator:      Winfried Neessen <doomy@pebcak.de>
#
# $Id$
#
# Last modified: [ 2011-01-05 00:14:04 ]

## This is the IperBackup::Config package {{{
package IperBackup::Config;

## Load some modules {{{
use warnings;
use strict;
use Carp qw( carp croak );
# }}}

## Defined constants {{{
use constant CONFFILE				=> './IperBackup.conf';				## Absolute path to config file if non is defined
use constant EXT_DEBUG				=> 0;						## Enable extended debug logging
use constant VERSION				=> '0.06';					## This modules version
# }}}

## Constuctor // new() {{{
sub new 
{

	## Read arguments
	my ( $class, %args ) = @_;

	## Get Logger object
	my $log = IperBackup::Main::get_logger( 'new' );

	## Reference object to class
	my $self = bless {}, $class;

	## Get the absolute path to the config file
	$self->{ 'conf_file' }	= delete( $args{ 'conf_file' } ) || CONFFILE;

	## Config file needs to be set for this object
	unless( defined( $self->{ 'conf_file' } ) )
	{
		$log->error( 'The config file path is mandatory for IperBackup::Config::new()' );
		return undef;
	}

	## Return the object
	return $self;

}
# }}}

## Read the configuration file // readconf() {{{
sub readconf
{

	## Get object
	my $self = shift;
	
	## Get Logger object
	my $log = IperBackup::Main::get_logger( 'readconf' );

	## Get filename of config file
	my $conffile = $self->is_present();

	## Die if the conffile is non-present
	unless( defined( $conffile ) ) 
	{
		$log->error( 'Config file ' . $self->{ 'conf_file' } . ' not present. Aborting.' );
		return undef;
	}

	## Open the config file
	open ( CONFIG, $conffile )
		or $log->logcroak( 'Unable to open config file: ' . $! );

	## Read the whole config file into an array
	my @config = <CONFIG>;

	## Close the file handle
	close( CONFIG );

	## Interpret the config values
	my $config = $self->process( @config );

	## Return the config hashref
	return $config;

}
# }}}

## Interpret the single config file values // process() {{{
sub process
{

	## Get object
	my ( $self, @config ) = @_;
	my $confhash;
	
	## Get Logger object
	my $log = IperBackup::Main::get_logger( 'process' );

	## Interpret the single config lines
	foreach my $line ( @config )
	{

		## Clean up line feeds
		chomp( $line );

		## Ignore comments and blank lines
		next if $line =~ /^(#|;)/;
		next if $line eq '';

		## Extract key and value
		$line =~ /\s*(\w+)\s*=\s*"(.+?)".*/i;
		$confhash->{ $1 } = $2;
		
		## Log something if ext. debug is enabled
		EXT_DEBUG && $log->debug( 
			'Received config line: ' . $line . ' // ',
			'Extracted -> Key: ' . $1 . ' -- Value: ' . $2
		);

	}

	## Return the hashref
	return $confhash;

}
# }}}

## Verify that the config file is present // is_present() {{{
sub is_present
{

	## Get object
	my $self = shift;
	
	## Get Logger object
	my $log = IperBackup::Main::get_logger( 'is_present' );

	## Set $conffile variable to the conf_file setting in the object
	my $conffile = $self->{ 'conf_file' } if -r $self->{ 'conf_file' };

	## Return the full path
	return $conffile;

}
# }}}


## Every module needs a true ending...
1;
# }}}
__END__
=head1 LICENSE
Copyright (c) 2010, Winfried Neessen <doomy@dokuleser.org>
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:
    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in the
      documentation and/or other materials provided with the distribution.
    * Neither the name of the neessen.net nor the
      names of its contributors may be used to endorse or promote products
      derived from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> BE LIABLE FOR ANY
DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


IperBackup/Update.pm
#!/usr/bin/perl -wt
#
# Filename:     IperBackup/Update.pm
# Description:  Update module to IperBackup
# Creator:      Winfried Neessen <doomy@pebcak.de>
#
# $Id$
#
# Last modified: [ 2010-12-13 15:55:50 ]

## This is the IperBackup::Update package {{{
package IperBackup::Update;

## Load some modules {{{
use warnings;
use strict;
use Carp qw( carp croak );
#use Data::Dumper;
use LWP::UserAgent;
use URI;
# }}}

## Defined constants {{{
use constant BASE_URL				=> 'http://blog.pebcak.de/tmp/IperBackupVersion.txt';		## URL with version sting of latest release
use constant DL_URL				=> 'http://nop.li/iperbackup';					## URL where to get latest release
use constant EXT_DEBUG				=> 0;								## Enable extended debug logging
use constant VERSION				=> '0.05';							## This modules version
# }}}

## Constuctor // new() {{{
sub new 
{

	## Read arguments
	my $class = shift;
	my $args = shift;

	## Get Logger object
	my $log = IperBackup::Main::get_logger( 'new' );

	## Reference object to class
	my $self = bless {}, $class;

	## Create LWP object
	$self->{ 'ua' } = LWP::UserAgent->new
	(

		agent		=> 'IperBackup/v' . VERSION,
		keep_alive	=> 1,
		env_proxy	=> 1,
		show_progress	=> 0,

	);

	## Return the object
	return $self;

}
# }}}

## Check version string of latest release // checkVersion() {{{
sub checkVersion
{

	## Get object and version of current release
	my ( $self, $version ) = @_;

	## Don't go further if no version has been given
	return undef unless( defined( $version ) );
	
	## Get Logger object
	my $log = IperBackup::Main::get_logger( 'checkVersion' );

	## Create URI object from uri string
	my $url = URI->new( BASE_URL );

	## Create a new HTTP request
	$self->{ 'request' } = $self->{ 'ua' }->get( $url );

	## Download is done... let's see if it was a success
	if( $self->{ 'request' }->is_success == 1 )
	{

		## Store latest version in object
		$self->{ 'latest' } = $self->{ 'request' }->decoded_content;
		chomp( $self->{ 'latest' } );

	} else {

		## The download was not successfull
		$log->warn( 'Couldn\'t fetch version information.' );

	}

	## Check if latest version is higher than current
	if( $self->{ 'latest' } > $version )
	{

		print "A new version of IperBackup is available. You are currently using v" . $version . ", but\n";
		print "the latest available version is v" . $self->{ 'latest' } . "\n\n";
		print "It is recommended to upgrade to the latest release. Find it at:\n";
		print DL_URL . "\n";
		my $foo = <STDIN>;

	}

	## Return to the caller
	return undef;

}
# }}}



## Every module needs a true ending...
1;
# }}}
__END__
=head1 LICENSE
Copyright (c) 2010, Winfried Neessen <doomy@dokuleser.org>
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:
    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in the
      documentation and/or other materials provided with the distribution.
    * Neither the name of the neessen.net nor the
      names of its contributors may be used to endorse or promote products
      derived from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> BE LIABLE FOR ANY
DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


IperBackup/Download.pm
#!/usr/bin/perl -wt
#
# Filename:     IperBackup/Download.pm
# Description:  Download module to IperBackup
# Creator:      Winfried Neessen <doomy@pebcak.de>
#
# $Id$
#
# Last modified: [ 2010-09-27 10:33:25 ]

## This is the IperBackup::Process package {{{
package IperBackup::Download;

## Load some modules {{{
use warnings;
use strict;
use Carp qw( carp croak );
#use Data::Dumper;
use LWP::UserAgent;
use Time::HiRes;
use URI;
# }}}

## Defined constants {{{
use constant EXT_DEBUG				=> 0;								## Enable extended debug logging
use constant VERSION				=> '0.100';							## This modules version
# }}}

## Constuctor // new() {{{
sub new 
{

	## Read arguments
	my $class = shift;
	my $args = shift;

	## Get Logger object
	my $log = IperBackup::Main::get_logger( 'new' );

	## Reference object to class
	my $self = bless {}, $class;

	## Create LWP object
	$self->{ 'ua' } = LWP::UserAgent->new
	(

		agent		=> 'IperBackup/v' . VERSION,
		keep_alive	=> 1,
		env_proxy	=> 1,
		show_progress	=> 1,

	);

	## Return the object
	return $self;

}
# }}}

## Download the file // download() {{{
sub download
{

	## Get object and URL to fetch
	my ( $self, $uri, $file ) = @_;

	## Don't go further if no URL or filename has been given
	return undef unless( defined( $uri ) and defined( $file ) );
	
	## Get Logger object
	my $log = IperBackup::Main::get_logger( 'download' );

	## Create URI object from uri string
	my $url = URI->new( $uri );

	## Log an information message
	$log->info( 'Downloading ' . $url . ' into ' . $file . '...' );
	
	## Start some benchmarking
	my $bm_start = [ Time::HiRes::gettimeofday ];

	## Create a new HTTP request
	$self->{ 'request' } = $self->{ 'ua' }->get( $url, ':content_file' => $file );

	## Download is done... let's see if it was a success
	if( $self->{ 'request' }->is_success == 1 )
	{

		## Finish the benchmarking
		$log->info( 'Download successfully finished in ' . sprintf( '%.3f', Time::HiRes::tv_interval( $bm_start ) ) . ' seconds' );

	} else {

		## The download was not successfull
		$log->error( 'Download unsuccessfully finished.' );

	}

	## Return to the caller
	return undef;

}
# }}}



## Every module needs a true ending...
1;
# }}}
__END__
=head1 LICENSE
Copyright (c) 2010, Winfried Neessen <doomy@dokuleser.org>
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:
    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in the
      documentation and/or other materials provided with the distribution.
    * Neither the name of the neessen.net nor the
      names of its contributors may be used to endorse or promote products
      derived from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> BE LIABLE FOR ANY
DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


IperBackup.pl
#!/usr/bin/perl
#
# Filename:     IperBackup.pl
# Description:  Main script of Ipernity::Backup
# Creator:      Winfried Neessen <doomy@pebcak.de>
#
# $Id$
#
# Last modified: [ 2011-01-13 13:47:12 ]

## This is the IperBackup::Main package {{{
package IperBackup::Main;

### Global modules {{{
use strict;
use warnings;
use Date::Manip;
use Getopt::Long;
use IperBackup::Config;
use IperBackup::Download;
use IperBackup::Process;
use IperBackup::Update;
use Ipernity::API;
use Log::Log4perl qw(:easy);
# }}}

### Basic configuration variables {{{
use constant DEFAULT_MEDIA				=> 'photo,video,audio,other';			## Default media types
use constant EXT_DEBUG					=> 0;						## Enable extended debug-logging
use constant LOGLEVEL					=> 'INFO';					## Set the log level
use constant OUTDIR					=> '/var/tmp';					## Default output directory
use constant VERSION					=> '0.09';					## Current version number
# }}}

## Define global variables {{{
my ( $config );
# }}}

### Log4Perl Configuration {{{
my $log4perl = q(
	log4perl.rootLogger					= ) . LOGLEVEL . q(, Screen
	log4perl.appender.Screen				= Log::Log4perl::Appender::Screen
	log4perl.appender.Screen.stderr				= 0
	log4perl.appender.Screen.layout				= PatternLayout
	log4perl.appender.Screen.layout.ConversionPattern	= %d [%p]: %m%n
);
Log::Log4perl->init( \$log4perl );
Log::Log4perl->wrapper_register(__PACKAGE__);
# }}}

### Check version of Ipernity::API module {{{
BEGIN
{

	my $API = $Ipernity::API::VERSION;
	do{ print "Ipernity::API v0.10 or higher required.\n"; exit 127; }
		unless( $API >= 0.10 );

}
# }}}

### Main subroutine // main() {{{
sub main
{

	## Get a log object
	my $log = get_logger('Main');

	## Check for updates
	checkUpdate();

	## Read command line arguments
	getArgs();
	
	## Create a config object {{{
	my $conf = IperBackup::Config->new
	(

		conf_file	=> $config->{ 'conffile' },

	);
	# }}}
		
	## Read the config file
	my $myconfig = $conf->readconf();

	## Don't execute anything if no config is present
	$log->logcroak( 'Unable to execute script, without a config file. Aborting.' ) unless defined( $myconfig );
	
	## Create an IperBackup::Download object
	my $dl = IperBackup::Download->new();
	
	## Create an API object {{{
	my $api = Ipernity::API->new
	(
		api_key		=> $myconfig->{ 'IPER_API_KEY' },
		secret		=> $myconfig->{ 'IPER_API_SECRET' },
		outputformat	=> $myconfig->{ 'IPER_API_OUTPUT' },

	);
	# }}}
	
	## Create an IperBackup::Process object {{{
	my $iper = IperBackup::Process->new
	(
		api		=> $api,
		config		=> $myconfig,
		media		=> $config->{ 'media' } || DEFAULT_MEDIA,
		tags		=> $config->{ 'tags' } || undef,
		startdate	=> $config->{ 'startdate' } || undef,
		enddate		=> $config->{ 'enddate' } || undef,
		nopermission	=> $config->{ 'nopermission' } || undef,
		timestamp	=> $config->{ 'timestamp' } || undef,
		album		=> $config->{ 'album' } || undef,

	);
	# }}}

	## An authtoken is mandatory, fetch one if non-existant {{{
	unless( defined( $myconfig->{ 'IPER_API_AUTHTOKEN' } ) and defined( $myconfig->{ 'IPER_USERID' } ) )
	{

		$iper->getToken( $api );

	}
	# }}}

	## Get number of documents
	my $docsnumber = $iper->getNumberDocs();

	## Show user how many docs going to be fetched
	$log->info( "Found " . $docsnumber . " documents in total for user account " . $iper->getUserInfo( 'username' ) . " (" . $iper->getUserInfo( 'realname' ) . ")" );

	## Get list of documents
	my $documents = $iper->getDocsList();

	## Calculate number of documents to fetch
	my $docCount = scalar( keys %{ $documents } );

	## Don't process any further if no document was found
	exit 127 if $docCount == 0;

	## Inform the user about the number of documents to be fetched
	$log->info( 'Will fetch ' . scalar( keys %{ $documents } ) . ' documents of the media type(s): ' . join( ', ', split( /,/, $config->{ 'media' } || DEFAULT_MEDIA ) ) );
	$log->info( 'Specified tags: ' . join( ', ', split( /,/, $config->{ 'tags' } ) ) ) if defined $config->{ 'tags' };

	## Decide which action to perform... the real download {{{
	if( defined( $config->{ 'download' } ) )
	{
		
		## Go through each document and download it
		foreach my $doc ( keys %{ $documents } )
		{

			## Make URL easily accessable
			my $url = $documents->{ $doc }->{ 'url' };

			## Generate absolute path to download filename
			my $file = $iper->isValidFile( $config->{ 'dir' } || OUTDIR, $documents->{ $doc }->{ 'fn' }, $doc );

			## Skip file if already present
			next if $file eq '___SKIP___';

			## If we wanna fetch comments, check if there are some
			if( defined( $config->{ 'comment' } ) )
			{

				## Get list of comments for the document ID
				my $commentPages = $iper->getCommentNumberPages( $doc );

				## Don't go any further if is no comment
				if( defined( $commentPages ) )
				{
					
					## Get comments via API
					my $comments = $iper->getCommentList( $doc, $commentPages );

					## Log a information message
					$log->info( 'Storing comments in ' .  $file . '_comments.txt' );

					## Open file to store comments in
					open( COMMENTS, '>', $file . '_comments.txt' )
						or $log->logcroak( 'Unable to open ' . $file . '_comments.txt for writing comments.' );

					## Go through each comment we have and store it
					foreach my $comid ( keys %{ $comments->{ $doc } } )
					{

						## Store the comment
						print COMMENTS "Comment ID:\t" . $comid . "\n";
						print COMMENTS "Comment URL:\t" . $comments->{ $doc }->{ $comid }->{ 'link' } . "\n";
						print COMMENTS "Written by:\t" . $comments->{ $doc }->{ $comid }->{ 'username' } . " (User ID: " . $comments->{ $doc }->{ $comid }->{ 'user_id' } . ")\n";
						print COMMENTS "Written on:\t" . scalar localtime( $comments->{ $doc }->{ $comid }->{ 'date' } ) . "\n";
						print COMMENTS "=" x 72 . "\n";
						print COMMENTS $comments->{ $doc }->{ $comid }->{ 'content' } . "\n";
						print COMMENTS "=" x 72 . "\n\n";

					}

					## Close comments file
					close( COMMENTS );

				}

			}

			## Download file (if commentsonly is not set)
			unless( defined( $config->{ 'commentsonly' } ) )
			{

				## Download file from Ipernity
				$dl->download( $url, $file );

				## Change date of downloaded file to creation date
				if( defined( $documents->{ $doc }->{ 'cdate' } ) and -f $file )
				{
			
					## Log a information message
					$log->info( 'Changing creation date of file to: ' . $documents->{ $doc }->{ 'cdate' } );

					## Convert date to unix timestamp
					my $cdate = UnixDate( $documents->{ $doc }->{ 'cdate' }, '%s' );

					## Update timestamp of downloaded file
					utime( $cdate, $cdate, $file );

				}

			}

		}

	}
	# }}}

	## ...or just creating a download list {{{
	elsif( defined( $config->{ 'list' } ) )
	{

		## Generate a full absolute path to the DL list
		my $file = $iper->isValidFile( $config->{ 'dir' } || OUTDIR, 'IperBackup.list', undef );

		## Inform the user where the DL list will be stored
		$log->info( 'Creating download link list: ' . $file );

		## Open the list a FH
		open( FH, '>', $file )
			or $log->logcroak( 'Unable to write DL list: ' . $! );
		
		## Go through each document and download it
		foreach my $doc ( keys %{ $documents } )
		{

			## Write it into the list
			print FH $documents->{ $doc }->{ 'url' } . "\n";

		}

		## Close the FH
		close( FH );

	}
	# }}}

}
# }}}

### Read arguments from command line // getArgs() {{{
sub getArgs
{

	GetOptions
	(

		'outdir|o=s'	=> \$config->{ 'dir' },
		'config|c=s'	=> \$config->{ 'conffile' },
		'help|h'	=> \$config->{ 'help' },
		'list|l'	=> \$config->{ 'list' },
		'download|d'	=> \$config->{ 'download' },
		'media|m=s'	=> \$config->{ 'media' },
		'tags|t=s'	=> \$config->{ 'tags' },
		'comments|n'	=> \$config->{ 'comment' },
		'commentsonly'	=> \$config->{ 'commentsonly' },
		'startdate|s=s'	=> \$config->{ 'startdate' },
		'enddate|e=s'	=> \$config->{ 'enddate' },
		'nopermission|p'=> \$config->{ 'nopermission' },
		'timestamp|ts'	=> \$config->{ 'timestamp' },
		'version|v'	=> \$config->{ 'version' },
		'album|a=s'	=> \$config->{ 'album' },

	);
	
	## Download mode is implied for commentsonly mode
	$config->{ 'download' } = 1 if( defined( $config->{ 'commentsonly' } ) );

	## Print version string if requested
	do { print "This is IperBackup v" . VERSION . " // (C) 2010-2011 Winfried Neessen\n"; exit 0 } if( defined( $config->{ 'version' } ) );

	## Check for conflicts
	showHelp() if( $config->{ 'help' } );
	showHelp() unless( defined( $config->{ 'list' } ) or defined( $config->{ 'download' } ) );
	showHelp() if( defined( $config->{ 'list' } ) and defined( $config->{ 'download' } ) );
	showHelp() if( defined( $config->{ 'list' } ) and defined( $config->{ 'comment' } ) );
	
	## Check media types
	if( defined( $config->{ 'media' } ) )
	{

		## Clean up whitespaces
		$config->{ 'media' } =~ s/\s//g;

		## Get array of types
		my @list = split( /,/, $config->{ 'media' } );
		showHelp() if( scalar( grep( !/\b(photo|audio|video|other)\b/, @list ) ) );

	}

	## Check tags
	if( defined( $config->{ 'tags' } ) )
	{

		## Clean up whitespaces
		$config->{ 'tags' } =~ s/,\s/,/g;

		## Count number of provided tags
		my @list = split( /,/, $config->{ 'tags' } );
		showHelp() if( $#list > 19 );

	}

}
# }}}

### Use IperBackup::Update to check for latest updates // checkUpdate() {{{
sub checkUpdate
{

	## Create an IperBackup::Update object
	my $update = IperBackup::Update->new();

	## Check if a new version is available
	$update->checkVersion( VERSION );

}
# }}}

### Provide some help messages if requested // showHelp() {{{
sub showHelp
{

	## Print message
	print "Usage: $0 [OPTIONS]\n";
	print "\n\t-a, --album\t\tSpecify an album ID to let IperBackup only fetch documents from that album";
	print "\n\t-c, --config\t\tSpecify absolute path to config file (Default: /etc/IperBackup.conf)";
	print "\n\t--comentsonly\t\tDownload comments only";
	print "\n\t-d, --download\t\tTell IperBackup to download all files in your account";
	print "\n\t-e, --enddate\t\tSpecify the maximum date of documents you are searching";
	print "\n\t-h, --help\t\tDisplay this help message.";
	print "\n\t-l, --list\t\tTell IperBackup to create a list of files in you account";
	print "\n\t-m, --media\t\tSpecify which media type to fetch. Possiblities are: audio, photo, other, video (Default: all)";
	print "\n\t-n, --comments\t\tFetch comments of the document if any (only works in download mode)";
	print "\n\t-o, --outdir\t\tSpecify absolute path to the output directory (Default: /var/tmp)";
	print "\n\t-p, --nopermission\tTell IperBackup to not add the permission type to the filename";
	print "\n\t-s, --startdate\t\tSpecify the minimum date of documents you are searching";
	print "\n\t-t, --tags\t\tForce IperBackup to fetch only files with a specific tag (max. 20 tags)";
	print "\n\t-ts, --timestamp\tIf a document is already present, IperBackup will add a timestamp and redownload the document";
	print "\n\t-v, --version\t\tShow version string";
	print "\n\n\n";

	## Exit with non-zero error code
	exit 127;

}
# }}}


### Execute main routine
main();
1;
__END__
=head1 LICENSE
Copyright (c) 2010, Winfried Neessen <doomy@dokuleser.org>
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:
    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in the
      documentation and/or other materials provided with the distribution.
    * Neither the name of the neessen.net nor the
      names of its contributors may be used to endorse or promote products
      derived from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> BE LIABLE FOR ANY
DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
